### PR TITLE
feat(commerce): post-interaction checkout flow

### LINKED ISSUE #728: Add seconds phase of the checkout flow (post-interaction)

### PR DESCRIPTION
**Pull request by raducristianpopa**
### Context

- closes #728

### CODE REVIEW

File: packages/commerce/backend/src/open-payments/service.ts [hunk 1]
```diff
@@ -155,9 +153,86 @@ export class OpenPayments implements IOpenPayments {
       throw new InternalServerError()
     })
 
+    await order.$query().patch({
+      quoteId: quote.id,
+      paymentPointerUrl: customerPaymentPointer.id,
+      continueToken: outgoingPaymentGrant.continue.access_token.value,
+      continueUri: outgoingPaymentGrant.continue.uri
+        .replace('localhost', 'rafiki-auth')
+        .replace('auth/', '')
```
Comment (raducristianpopa) [inline packages/commerce/backend/src/open-payments/service.ts:162]:
```suggestion
        .replace('auth/', '')
```

I think this is a mistake coming from Rafiki's auth. Waiting for Nathan's response.

File: packages/commerce/backend/src/open-payments/service.ts [hunk 2]
```diff
@@ -155,9 +153,85 @@ export class OpenPayments implements IOpenPayments {
...
+          {
+            accessToken: order.continueToken,
+            url: order.continueUri
+          },
+          {
+            interact_ref: interactRef
+          }
+        )
+        .catch((err) => {
+          this.logger.error('Could not finish the continuation request')
```
Comment (Tymmmy) [inline packages/commerce/backend/src/open-payments/service.ts:201]:
```suggestion
          this.logger.error('Could not finish the continuation request.')
```

File: packages/commerce/frontend/src/app/checkout/components/loader.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+      <motion.div
+        className="h-24 w-24 rounded-full border-[16px] border-gray-200 border-t-turqoise"
+        animate={{ rotate: 360 }}
+        transition={{
+          repeat: Infinity,
+          ease: 'linear',
+          duration: 1
+        }}
+      />
+      <h2 className="text-2xl uppercase">Processing ...</h2>
```
Comment (Tymmmy) [inline packages/commerce/frontend/src/app/checkout/components/loader.tsx:15]:
```suggestion
      <h2 className="text-2xl uppercase">Processing...</h2>
```

**Review by Tymmmy — APPROVED** [review]

**Review by Tymmmy — DISMISSED** [review]

File: packages/commerce/backend/src/open-payments/service.ts [hunk 4]
```diff
@@ -155,9 +153,85 @@ export class OpenPayments implements IOpenPayments {
       throw new InternalServerError()
     })
 
+    await order.$query().patch({
+      quoteId: quote.id,
+      paymentPointerUrl: customerPaymentPointer.id,
+      continueToken: outgoingPaymentGrant.continue.access_token.value,
+      continueUri: outgoingPaymentGrant.continue.uri
+        .replace('localhost', 'rafiki-auth')
```
Comment (dragosp1011) [inline packages/commerce/backend/src/open-payments/service.ts:161]: is this going to work on deployed version? should here be some env variables instead?

File: packages/commerce/backend/src/open-payments/service.ts [hunk 5]
```diff
@@ -155,9 +153,85 @@ export class OpenPayments implements IOpenPayments {
...
+            accessToken: continuation.access_token.value
+          },
+          {
+            quoteId: order.quoteId,
+            metadata: {
+              orderRef: order.id
+            }
+          }
+        )
+        .catch(() => {
```
Comment (dragosp1011) [inline packages/commerce/backend/src/open-payments/service.ts:219]: log the error as well

File: packages/commerce/backend/src/order/controller.ts [hunk 6]
```diff
@@ -80,4 +81,38 @@ export class OrderController implements IOrderController {
...
+        req.body
+      )
+
+      const order = await this.orderService.ensurePendingState(orderId)
+
+      if (result) {
+        const status = result === 'grant_rejected' ? 200 : 400
+        this.orderService.reject(orderId)
+        res
+          .status(result === 'grant_rejected' ? 200 : 400)
```
Comment (dragosp1011) [inline packages/commerce/backend/src/order/controller.ts:105]: use the status const

File: packages/commerce/backend/src/order/controller.ts [hunk 7]
```diff
@@ -80,4 +81,38 @@ export class OrderController implements IOrderController {
...
+          .json({ success: status === 200 ? true : false, message: 'SUCCESS' })
```
Comment (dragosp1011) [inline packages/commerce/backend/src/order/controller.ts:106]: message `SUCCESS` in both cases?

File: packages/commerce/backend/src/open-payments/service.ts (hunk 4 above)
Comment (raducristianpopa) [reply packages/commerce/backend/src/open-payments/service.ts:161]: I'll add a check to only replace it in a dev environment. 👍

File: packages/commerce/backend/src/open-payments/service.ts (hunk 5 above)
Comment (raducristianpopa) [reply packages/commerce/backend/src/open-payments/service.ts:219]: Open Payments already has a logger implemented which will automatically log the error. I'll need to remove `console.log(err)` from `catch` blocks.

**Review by Tymmmy — APPROVED** [review]

File: packages/commerce/backend/src/app.ts [hunk 8]
```diff
@@ -119,4 +120,26 @@ export class App {
...
     }
   }
+
+  private async processPendingOrders() {
+    const orderService = this.container.resolve('orderService')
+    const logger = this.container.resolve('logger')
+    return orderService
+      .processPendingOrders()
+      .catch(() => {
+        logger.error('Error while trying to process pending orders')
```
Comment (dragosp1011) [inline packages/commerce/backend/src/app.ts:130]: log the error as well

File: packages/commerce/backend/src/open-payments/service.ts [hunk 9]
```diff
@@ -59,8 +59,10 @@ export class OpenPayments implements IOpenPayments {
       .get({
         url: paymentPointerUrl
       })
-      .catch((err) => {
-        this.logger.error(err)
+      .catch(() => {
```
Comment (dragosp1011) [inline packages/commerce/backend/src/open-payments/service.ts:62]: `preparePayment`  method is way to big, can it be divided somehow?

Comment (raducristianpopa) [reply packages/commerce/backend/src/open-payments/service.ts:62]: Around 100 LOC seems reasonable to me. I don't see how we can abstract more methods. For example, I don't think moving `this.opClient.incomingPayment.create` (around 20 LOC) into a separate method would make a difference.

File: packages/commerce/backend/src/open-payments/service.ts [hunk 10]
```diff
@@ -215,25 +237,84 @@ export class OpenPayments implements IOpenPayments {
     return grant
   }
 
-  private async createQuote({
-    paymentPointerUrl,
+  private async getPaymentPointer(url: string) {
+    const paymentPointer = await this.opClient.paymentPointer
+      .get({
+        url: url
```
Comment (dragosp1011) [inline packages/commerce/backend/src/open-payments/service.ts:243]: `url` is enough here

File: packages/commerce/backend/src/order/controller.ts [hunk 11]
```diff
@@ -80,4 +81,39 @@ export class OrderController implements IOrderController {
...
+      // https://rafiki.dev/concepts/open-payments/grant-interaction/#endpoints
+      const { interactRef, result } = await validate(
+        finishOrderSchema,
+        req.body
+      )
+
+      const order = await this.orderService.ensurePendingState(orderId)
+
+      if (result) {
+        const status = result === 'grant_rejected' ? 200 : 400
```
Comment (dragosp1011) [inline packages/commerce/backend/src/order/controller.ts:102]: introducing a separate const for rejected state may make this a bit more readable
```
        const isRejected = result === 'grant_rejected'
        const status = isRejected ? 200 : 400
        const message = isRejected ? 'SUCCESS' : 'FAILED'
        this.orderService.reject(orderId)
        res
          .status(status)
          .json({ success: isRejected, message })

File: packages/commerce/frontend/src/app/checkout/confirmation.tsx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      hash,
+      interactRef: interact_ref,
+      result
+    })
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  if (data) {
+    let color: string = 'text-turqoise'
+    let text: string | undefined = 'Thank you for your order!'
```
Comment (dragosp1011) [inline packages/commerce/frontend/src/app/checkout/confirmation.tsx:31]: when can this be undefined?

**Review by dragosp1011 — APPROVED** [review]

**Pull request merged**